<?xml version="1.0"?>
<robot xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor"
       xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller"
       xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface">

  <property name="M_PI" value="3.1415926535897931" />
  <property name="VELOCITY_LIMIT_SCALE" value="0.6" />

  <property name="shoulder_lift_length" value="0.10" />
  <property name="shoulder_lift_radius" value="0.12" />

  <!-- for non-symmetric upper arm -->
  <property name="uar_xyz1" value= "0.21398" />
  <property name="uar_xyz2" value="-0.01621" />
  <property name="uar_xyz3" value="-0.00020" />
  <property name="uar_ixx"  value= "0.01537748957" />
  <property name="uar_ixy"  value= "0.00375711247" />
  <property name="uar_ixz"  value="-0.00070852914" />
  <property name="uar_iyy"  value= "0.07473670440" />
  <property name="uar_iyz"  value="-0.00017936450" />
  <property name="uar_izz"  value= "0.07608763307" />

  <property name="ual_xyz1" value= "0.21405" />
  <property name="ual_xyz2" value= "0.01658" />
  <property name="ual_xyz3" value="-0.00057" />
  <property name="ual_ixx"  value= "0.01530603856" />
  <property name="ual_ixy"  value="-0.00339324862" />
  <property name="ual_ixz"  value= "0.00060765455" />
  <property name="ual_iyy"  value= "0.07473694455" />
  <property name="ual_iyz"  value="-0.00019953729" />
  <property name="ual_izz"  value= "0.07601594191" />


  <macro name="m3_l_shoulder" params="side parent reflect *origin">

    <!-- Shoulder pan -->
    <joint name="left_arm_j0" type="revolute">
      <axis xyz="0 -1 0" />
      <insert_block name="origin" /> <!-- transform from parent link to this joint frame -->
      <parent link="${parent}"/>
      <child link="shoulderupper_LEFT"/>
      <limit effort="40" velocity="${30.0*M_PI/180}" lower="${-80*M_PI/180}" upper="${200*M_PI/180}" />
      <dynamics damping="1.0"/>
    </joint>
    <link name="shoulderupper_LEFT">
      <inertial>
        <mass value="2.5799322" />
        <origin xyz="-0.001201 0.024513 -0.098231" rpy="0 0 0"/>
        <inertia ixx="0.866179142480"  ixy="-0.06086507933"  ixz="-0.12118061183"
                 iyy="0.87421714893"  iyz="-0.05886609911"
                 izz="0.27353821674" />
      </inertial>

      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="${M_PI} 0 0 " />
	      <geometry name="r_shoulder_geom">
		      <mesh filename="package://${MESH}/A2R3/A2R3_RT1_shoulderRoll.dae" />
	      </geometry>  
      </visual>

      <collision>
	      <origin xyz="0 0 0" rpy="${M_PI} 0.0 0.0 " />
	      <geometry name="r_shoulder_collision_geom">
		      <mesh filename="package://${MESH_COLL}/A2R3/A2R3_RT1_shoulderRoll.stl" />
	      </geometry>
      </collision>
    </link>

    
    <joint name="left_arm_j1" type="revolute">
      <axis xyz="-1 0 0"/>
     <limit effort="40" velocity="${30*M_PI/180}" upper="${24*M_PI/180}" lower="${-150*M_PI/180}" />    
      <origin xyz="0 0 0" rpy="0 0 0" />
      <parent link="shoulderupper_LEFT"/>
      <child link="shoulderpitch_LEFT"/>
      <dynamics damping="1.0"/>
    </joint>
    <link name="shoulderpitch_LEFT">

      <inertial>
        <mass value="2.74988" />
        <origin xyz="0.02195 -0.02664 -0.03127" rpy="0 0 0"/>
        <inertia ixx="0.02105584615" ixy="0.00496704022"  ixz="-0.00194808955"
                 iyy="0.02127223737" iyz="0.00110425490"
                 izz="0.01975753814" />
      </inertial>

      <visual>
	      <material name="GreyB" />
	      <origin xyz="0 0 0" rpy="0 0 ${M_PI}" />
	      <geometry name="shoulderpitch_LEFT">
		      <mesh filename="package://${MESH}/A2R3/A2R3_RT2_shoulderPitch.dae" />		      
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 ${M_PI}" />
	      <geometry name="shoulderpitch_LEFT_collision">
		      <mesh filename="package://${MESH_COLL}/A2R3/A2R3_RT2_shoulderPitch.stl" />
	      </geometry>
      </collision>
    </link>

  

    <joint name="left_arm_j2" type="revolute">
      <axis xyz="0 0 -1" />
      <origin xyz="0 0.03175 -0.27857" rpy="0 0 0" />
      <parent link="shoulderpitch_LEFT" />
      <child link="bicep_LEFT"/>
      <limit effort="20" velocity="${50*M_PI/180}" lower="${-85*M_PI/180}" upper="${85*M_PI/180}" />    
      <dynamics damping="1.0"/>
    </joint>
    <link name="bicep_LEFT">
      <inertial>        
        <mass value="2.296" />
        <origin xyz="-0.005767 0.084539 0.001385"/>
	<inertia ixx="0.013031"
	iyy="0.002070"
	izz="0.013173"
	ixy="-0.000564"
	ixz="-0.000077"
	iyz="0.000162" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="bicep_LEFT_geom">
		      <mesh filename="package://${MESH}/A2R3/A2R3_LT3_bicep.dae" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="bicep_LEFT_geom_collision">
		      <mesh filename="package://${MESH_COLL}/A2R3/A2R3_LT3_bicep.stl" />
	      </geometry>
      </collision>
    </link>

  </macro>



  <!-- ============================   Upper Arm   ============================ -->


  <macro name="m3_l_upper_arm" params="side parent reflect">

	  <joint name="left_arm_j3" type="revolute">
		  <axis xyz="0 -1 0" />
		  <origin xyz="0.00502 0  0" rpy="0 0 0" />
      <parent link="${parent}"/>
      <child link="elbowclevis_LEFT"/>
	<limit effort="20" velocity="${50*M_PI/180}" lower="${-0.278978*M_PI/180}" upper="${132.8175*M_PI/180}" />    
    </joint>
    <link name="elbowclevis_LEFT">

      <inertial>
        <!-- NOTE:reflect==-1 for LEFT side, reflect==1 for the left side -->
        <mass value="6.01769" />
        <origin xyz="${0.5*((reflect+1)*ual_xyz1+(1-reflect)*uar_xyz1)} ${0.5*((reflect+1)*ual_xyz2+(1-reflect)*uar_xyz2)} ${0.5*((reflect+1)*ual_xyz3+(1-reflect)*uar_xyz3)}" />
        <inertia  ixx="${0.5*((reflect+1)*ual_ixx+(1-reflect)*uar_ixx)}"
                  ixy="${0.5*((reflect+1)*ual_ixy+(1-reflect)*uar_ixy)}"
                  ixz="${0.5*((reflect+1)*ual_ixz+(1-reflect)*uar_ixz)}"
                  iyy="${0.5*((reflect+1)*ual_iyy+(1-reflect)*uar_iyy)}"
                  iyz="${0.5*((reflect+1)*ual_iyz+(1-reflect)*uar_iyz)}"
                  izz="${0.5*((reflect+1)*ual_izz+(1-reflect)*uar_izz)}" />
      </inertial>

      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 ${M_PI}" />
	      <geometry name="elbowclevis_LEFT_geom">
		      <mesh filename="package://${MESH}/A2R3/A2R3_RT4_elbow.dae" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 ${M_PI}" />
	      <geometry name="elbowclevis_LEFT_collision">
		      <mesh filename="package://${MESH_COLL}/A2R3/A2R3_RT4_elbow.stl" />
	      </geometry>
      </collision>

    </link>

  

    <joint name="left_arm_j4" type="revolute">
	    <axis xyz="0 0 -1" />
      <limit effort="4" velocity="${150*M_PI/180}" lower="${-190.7*M_PI/180}" upper="${19.7*M_PI/180}" /> 
      <origin xyz="0 0 -0.27747" rpy="0 0 0" />
      <parent link="elbowclevis_LEFT"/>
      <child link="wrist_LEFT"/>
      <dynamics damping="1.0"/>
    </joint>
    <link name="wrist_LEFT">

      <inertial>
        <mass value="1.90327" />
        <origin xyz="0.01014 0.00032 -0.01211" />
        <inertia  ixx="0.00346541989" ixy="0.00004066825"  ixz="0.00043171614"
                  iyy="0.00441606455"  iyz="-0.00003968914"
                  izz="0.00359156824" />
      </inertial>

      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="wrist_LEFT_geom">
		      <mesh filename="package://${MESH}/A2R3/A2R3_T5_forearmRoll.dae" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="wrist_LEFT_collision">
		      <mesh filename="package://${MESH_COLL}/A2R3/A2R3_T5_forearmRoll.stl" />
	      </geometry>
      </collision>
    </link>
       

  </macro>


  <!-- ============================   Forearm   ============================ -->

  <macro name="m3_l_forearm" params="side parent reflect">

    <joint name="left_arm_j5" type="revolute">
      <dynamics damping="1.0" />
		  <axis xyz="0 -1 0" />
      <origin xyz="0 0 0" rpy="0 0 0" />
      <parent link="${parent}"/>
      <child link="w_differential_LEFT"/>
	<limit effort="8" velocity="${40*M_PI/180}" lower="${-58.83*M_PI/180}" upper="${58.83*M_PI/180}" /> 
      <dynamics damping="1.0"/>
    </joint>
    <link name="w_differential_LEFT">

      <inertial>
        <mass value="2.57968" />
        <origin xyz="0.18791 -0.00017 -0.00912" />
        <inertia ixx="0.00364857222" ixy="0.00005215877" ixz="0.00071534842"
                 iyy="0.01507736897" iyz="-0.00001310770"
                 izz="0.01659310749" />
      </inertial>

      <visual>
	      <material name="GreyB" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="w_differential_LEFT_geom">
		      <mesh filename="package://${MESH}/A2R3/A2R3_T6_wristDiff.dae" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="w_differential_LEFT_collision">
		      <box size="0.01 0.01 0.01" />
	      </geometry>
      </collision>
    </link>

    

    <joint name="left_arm_j6" type="revolute">
	    <axis xyz="-1 0 0" />
      <limit effort="8" velocity="${40*M_PI/180}" lower="${-60*M_PI/180}" upper="${60*M_PI/180}" /> 
      <origin xyz="0 0 0" rpy="0 0 0" />
      <parent link="w_differential_LEFT"/>
      <child link="handmount_LEFT"/>
      <dynamics damping="0.1"/>
    </joint>
    <link name="handmount_LEFT">

      <inertial>
        <mass value="0.61402" />
        <origin xyz="-0.00157 0.0 -0.00075" />
        <inertia ixx="0.00065165722" ixy="0.00000028864" ixz="0.00000303477"
                 iyy="0.00019824443" iyz="-0.00000022645"
                 izz="0.00064450498" />
      </inertial>

      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="handmount_LEFT_geom">
		       <mesh filename="package://${MESH}/A2R3/A2R3_T7_wristYaw.dae" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="handmount_LEFT_collision">
		      <mesh filename="package://${MESH_COLL}/A2R3/A2R3_T7_wristYaw.stl" />
	      </geometry>
      </collision>
    </link>

  </macro>



  <macro name="m3_l_arm" params="side parent reflect *origin">
    <m3_l_shoulder side="${side}" reflect="${reflect}" parent="${parent}">
      <insert_block name="origin" />
    </m3_l_shoulder>
    <m3_l_upper_arm side="${side}" reflect="${reflect}" parent="bicep_LEFT" />
    <m3_l_forearm side="${side}" reflect="${reflect}" parent="wrist_LEFT" />
  </macro>


</robot>
